### PR TITLE
Detect N+1 hydration and throw an exception

### PR DESCRIPTION
**Pull request by qnkhuat**
Throw an exception if a n+1 hydration is detected. Only do this during development and testing.

```clojure

(t2/hydrate (t2/select :model/Database {:limit 1}) :tables)
;; => this is fine

(t2/hydrate (t2/select :model/Database {:limit 2}) :tables)
;; => Execution error (ExceptionInfo) at metabase.models.interface/hydrate-with-strategy-before-method-toucan2-tools-hydrate-multimethod-simple (REPL:675).
;;    N+1 hydration detected!!! Model :model/Database, hydration key :tables
```

### CODE REVIEW

**Comment by piranha** [comment]
Not sure what happens, but everything's broken 😁

**Review by sansolovyov — APPROVED** [review]
Doing god's work

**Review by tsmacdonald — APPROVED** [review]
I'm amazed this passes CI, but :clap:

File: src/metabase/models/interface.clj [hunk 1]
```diff
@@ -704,6 +705,26 @@
   [_original-model dest-key _hydrated-key]
   [(u/->snake_case_en (keyword (str (name dest-key) "_id")))])
 
+(methodical/defmethod t2.hydrate/hydrate-with-strategy :around ::t2.hydrate/multimethod-simple
+  "Throws an error if do simple hydrations that make DB call on a sequence."
+  [model strategy k instances]
+  (if (or config/is-prod? (< (count instances) 2)
```
Comment (tsmacdonald) [inline src/metabase/models/interface.clj:711]:
```suggestion
  (if (or config/is-prod?
          (< (count instances) 2)
```

nit: weird to have two conditions on one line and the third on another, I'd do one line per condition

**Comment by tsmacdonald** [comment]
image
did these commits get into master via a separate PR or something?

**Comment by qnkhuat** [comment]
@tsmacdonald yes, I didn't expect to get the CI passes on this so I moved those commit to its own PR https://github.com/metabase/metabase/pull/40345

**Comment by tsmacdonald** [comment]
Awesome, thanks Ngoc!

**Pull request merged**